remotes/certutil/certutil.go

```diff
@@ -0,0 +1,106 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package certutil
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// SystemCertPool returns a copy of the system cert pool,
+// returns an error if failed to load or empty pool on windows.
+//
+// SystemCertPool was ported from Docker 19.03
+// https://github.com/docker/engine/blob/v19.03.1/vendor/github.com/docker/go-connections/tlsconfig/certpool_go17.go#L12
+func SystemCertPool() (*x509.CertPool, error) {
+	certpool, err := x509.SystemCertPool()
+	if err != nil && runtime.GOOS == "windows" {
+		return x509.NewCertPool(), nil
+	}
+	return certpool, err
+}
+
+func hasFile(files []os.FileInfo, name string) bool {
+	for _, f := range files {
+		if f.Name() == name {
+			return true
+		}
+	}
+	return false
+}
+
+// ReadHostCertsDirectory reads the directory for TLS certificates
+// including roots and certificate pairs and updates the
+// provided TLS configuration.
+//
+// ReadHostCertsDirectory was ported from Docker 19.03 and compatible with
+// /etc/docker/certs.d/<host> directory layout.
+//
+// https://github.com/docker/engine/blob/v19.03.1/registry/registry.go#L57
+func ReadHostCertsDirectory(tlsConfig *tls.Config, directory string) error {
+	fs, err := ioutil.ReadDir(directory)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	for _, f := range fs {
+		if strings.HasSuffix(f.Name(), ".crt") {
```

Comment:
This is way too Docker specific and not something we want to replicate in containerd. There is no standard here around these extensions mapping to specific types of file. As an alternative, a function which takes in a glob pattern for loading CAs or other types is reasonable.

Reply:
What would glob look like? How will it be configured for cri plugin? (and buildkit)

Reply:
ping

Reply:
You could load certs by `/etc/containerd/<hostname>/*.crt` if that is the pattern. I would also separate adding of the CAs from adding the client certs, not necessary to be in the same function.

Reply:
updated